yoyodyne/models/pointer_generator.py

```diff
@@ -499,31 +292,78 @@ def forward(self, batch: batches.PaddedBatch) -> torch.Tensor:
+                self.source_encoder.num_directions,
+            )
         else:
-            predictions = self.decode(
-                len(batch),
-                (h_0, c_0),
-                batch.source.padded,
-                source_encoded,
-                batch.source.mask,
-                features_encoded,
-                batch.features.mask,
-                batch.target.padded,
+            last_hiddens = self.init_hiddens(
+                len(batch), self.source_encoder.layers
             )
+        if not self.has_feature_encoder:
+            if self.beam_width is not None and self.beam_width > 1:
+                # predictions = self.beam_decode(
+                # batch_size, x_mask, encoder_out, beam_width=self.beam_width
+                # )
+                raise NotImplementedError
+            else:
+                predictions = self.decode(
+                    batch.source.padded,
+                    encoded,
+                    batch.source.mask,
+                    target=batch.target.padded,
+                    last_hiddens=last_hiddens,
+                )
+        else:
+            assert batch.has_features
+            features_encoder_output = self.feature_encoder(batch.features)
+            features_encoded = features_encoder_output.encoded
+            if features_encoder_output.has_hiddens:
+                h_features, c_features = features_encoder_output.hiddens
+                h_features, c_features = self._reshape_hiddens(
+                    h_features,
+                    c_features,
+                    self.feature_encoder.layers,
+                    self.feature_encoder.num_directions,
+                )
+            else:
+                h_features, c_features = self.init_hiddens(
+                    len(batch), self.source_encoder.layers
+                )
+            # Joins hidden states.
+            h_source, c_source = last_hiddens
+            h0 = self.merge_h(torch.cat([h_source, h_features], dim=2))
+            c0 = self.merge_c(torch.cat([c_source, c_features], dim=2))
+            last_hiddens = h0, c0
+            if self.beam_width is not None and self.beam_width > 1:
+                # predictions = self.beam_decode(
+                #   batch_size, x_mask, encoder_out,
+                #   beam_width=self.beam_width
+                # )
+                raise NotImplementedError
+            else:
+                predictions = self.decode(
+                    batch.source.padded,
+                    encoded,
+                    batch.source.mask,
+                    feature_enc=features_encoded,
+                    feature_mask=batch.features.mask,
+                    target=batch.target.padded,
+                    last_hiddens=last_hiddens,
+                )
+
         # -> B x seq_len x output_size.
         predictions = predictions.transpose(0, 1)
         return predictions
+
+    @staticmethod
+    def _reshape_hiddens(H, C, layers, num_directions):
```

Comment:
I think capitalization is more conventional

Reply:
PEP-8 says otherwise. (I am not completely opposed to exceptions when you're matching an explicit and difficult function, say one that contrasts `H` and `h`, but that's not applicable here.) I fixed it upstream since it's not directly related to your PR here anyways...sorry for the drive-by.